src/realm/sync/subscriptions.hpp

```diff
@@ -0,0 +1,287 @@
+        }
+
+        // used in tests.
+        inline friend std::ostream& operator<<(std::ostream& o, const iterator& it)
+        {
+            o << "SubscriptionSet::iterator(" << std::hex << it.m_parent << ", " << std::dec << it.m_sub_it.index()
+              << ")";
+            return o;
+        }
+
+        iterator& operator++();
+        iterator operator++(int);
+
+    protected:
+        friend class SubscriptionSet;
+
+        iterator(const SubscriptionSet* parent, LnkLst::iterator it);
+
+    private:
+        const SubscriptionSet* m_parent;
+        LnkLst::iterator m_sub_it;
+        mutable Subscription m_cur_sub;
+    };
+
+    using const_iterator = const iterator;
+
+    // This will make a copy of this subscription set with the next available version number and return it as
+    // a mutable SubscriptionSet to be updated. The new SubscriptionSet's state will be Uncommitted. This
+    // subscription set will be unchanged.
+    SubscriptionSet make_mutable_copy() const;
+
+    // The query version number used in the sync wire protocol to identify this subscription set to the server.
+    int64_t version() const;
+
+    // The current state of this subscription set
+    State state() const;
+
+    // The error string for this subscription set if any.
+    StringData error_str() const;
+
+    // Returns the number of subscriptions in the set.
+    size_t size() const;
+
+    // A const_iterator interface for finding/working with individual subscriptions.
+    const_iterator begin() const;
+    const_iterator end() const;
+
+    // Returns a const_iterator to the query matching either the name or Query object, or end() if no such
+    // subscription exists.
+    const_iterator find(StringData name) const;
+    const_iterator find(const Query& query) const;
+
+    // A mutable iterator interface for working with individual subscriptions in when the set is in the
+    // Uncommitted state.
+    iterator begin();
+    iterator end();
+
+    // Erases a subscription pointed to by an iterator. Returns the "next" iterator in the set - to provide
+    // STL compatibility. The SubscriptionSet must be in the Uncommitted state to call this - otherwise
+    // this will throw.
+    iterator erase(iterator it);
+
+    // Erases all subscriptions in the subscription set.
+    void clear();
+
+    // Inserts a new subscription into the set if one does not exist already - returns an iterator to the
+    // subscription and a bool that is true if a new subscription was actually created. The SubscriptionSet
+    // must be in the Uncommitted state to call this - otherwise this will throw.
+    //
+    // The Query portion of the subscription is mutable, however the name portion is immutable after the
+    // subscription is inserted.
+    //
+    // If insert is called twice for the same query, this is a no-op.
+    std::pair<iterator, bool> insert(const Query& query, util::Optional<std::string> name = util::none);
```

Comment:
This should have a `bool update = false` argument to allow updating queries by name.

Reply:
I mean, then it wouldn't be `insert`, it'd be update? Can't the SDKs just do
```
if (auto it = set.find(name); it != set.end()) {
    it->update(new_query);
}
```

I tried to write this so it'd be idiomatic C++ that behaves the way normal STL containers behave.

Reply:
I see. We can, but then it means all SDKs will need to do the same thing. I'd generally prefer if the Core API is reasonably close to the SDK API as that leaves less room for interpretation, but if you'd rather keep the C++ API more idiomatic, I don't have strong objections.

Reply:
I've updated this to hopefully be what you had in mind. Instead of an insert method, there's an insert_or_assign method that either creates a new subscription or updates it the way the `update_query` method worked on `Subscription`. There's both an `insert_or_assign(std::string_view name, const Query&)` and an `insert_or_assign(const Query&)` for anonymous subscriptions. Let me know if this is closer to what you had in mind.

Reply:
Hm, I may have not been clear enough - what I was thinking of in terms of the API would be something like (excuse my pseudocode C++):

```cpp
std::pair<iterator, bool> insert(const Query& query, std::string_view name = "", allow_update = false) {
  // Can also be util::Optional
  if (!name) {
	name = query.description();
  }

  if (auto it = set.find(name); it != set.end()) {
    if (allow_update) {
      it->update(new_query);
    } else if (query.description() != it->query_string() {
      throw std::runtime_error(util::format("Can't update a query with name $0 from $1 to $2", ...);
    }
  } else {
    // Create a new sub
  }
}
```

Reply:
Maybe I'm not being clear - I am not a fan of that API and I'm trying to find a compromise. Boolean flags with a default value on a public API raises an eyebrow, and having a default boolean flag that controls whether the function you're calling will throw or not raises both eyebrows and furrows the brow.

I like the idea of removing the update method from the Subscription object and treating SubscriptionSet as a set that you can insert/assign (update) unique values into. I don't think the insert method should throw though unless an error has occurred, and this doesn't feel like an error condition - more like using exceptions for flow control. 

If we have methods to test whether a query or name is in the set and a way to insert/assign into the set, is that not sufficient?

Reply:
My suggestion mirrors fairly closely the SDK public API. As mentioned, if that would make a weird C++ API, I don't mind implementing that logic at the SDK layer, I just thought it would be more efficient to have it one place, but the code is so trivial to write we'll probably spend more time discussing it than actually implementing it. The current `insert_or_assign` approach is sufficient to do that.